### ISSUE TITLE
[JENKINS-45575] Multiple assignment in pipeline fails with LHS Error

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/workflow-cps-plugin#1337]
<p>Multiple assignment like:</p>

<pre>
def (a, b) = [1,2]</pre>
<p> No longer works, I believe this worked in Jenkins around 2.65 (and the pipeline components from around that release)</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kbrowder.jpg" title="kbrowder's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kbrowder">kbrowder</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-45575" target="_blank">Multiple assignment in pipeline fails with LHS Error</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer">abayer</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>component(s)</b>: workflow-cps-plugin
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2018-03-07T18:34:27+00:00
<li><b>votes</b>: 7
<li><b>watchers</b>: 16
<li><b>imported</b>: 2025-12-07
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Multiple assignment like:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def (a, b) = [1,2]</pre>
</div></div>
<p> No longer works, I believe this worked in Jenkins around 2.65 (and the pipeline components from around that release)</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Jenkins 2.69</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/duemir.png" title="duemir's avatar" /> duemir:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-45575?focusedId=322822&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-322822">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> There is also an issue with a simple instantiation of the `groovy.lang.Tuple`. It seems to be blocked by the Script Security. Details are below. Let me know if this is a separate issue and I will raise a different ticket.</p>

<p>Jenkins 2.95
Script Security 1.39</p>

<p>Test Pipeline</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">node {
    def output = stuff()
    echo output[0]
    echo output[1]
}

def stuff() {
    <span class="code-keyword">return</span> <span class="code-keyword">new</span> Tuple(1, <span class="code-quote">'banana'</span>)
}
</pre>
</div></div>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">org.jenkinsci.plugins.scriptsecurity.sandbox.RejectedAccessException: unclassified <span class="code-keyword">new</span> groovy.lang.Tuple java.lang.<span class="code-object">Integer</span> java.lang.<span class="code-object">String</span>
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onNewInstance(SandboxInterceptor.java:144)
	at org.kohsuke.groovy.sandbox.impl.Checker$3.call(Checker.java:197)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedConstructor(Checker.java:202)
	at com.cloudbees.groovy.cps.sandbox.SandboxInvoker.constructorCall(SandboxInvoker.java:21)
	at WorkflowScript.stuff(WorkflowScript:8)
	at WorkflowScript.run(WorkflowScript:2)
	at ___cps.transform___(Native Method)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:96)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.fixArg(FunctionCallBlock.java:82)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
	at com.cloudbees.groovy.cps.impl.ConstantBlock.eval(ConstantBlock.java:21)
	at com.cloudbees.groovy.cps.Next.step(Next.java:83)
	at com.cloudbees.groovy.cps.Continuable$1.call(Continuable.java:174)
	at com.cloudbees.groovy.cps.Continuable$1.call(Continuable.java:163)
	at org.codehaus.groovy.runtime.GroovyCategorySupport$ThreadCategoryInfo.use(GroovyCategorySupport.java:122)
	at org.codehaus.groovy.runtime.GroovyCategorySupport.use(GroovyCategorySupport.java:261)
	at com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:163)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.access$001(SandboxContinuable.java:19)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable$1.call(SandboxContinuable.java:35)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable$1.call(SandboxContinuable.java:32)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.GroovySandbox.runInSandbox(GroovySandbox.java:108)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.run0(SandboxContinuable.java:32)
	at org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:174)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:331)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$100(CpsThreadGroup.java:82)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:243)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:231)
	at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:64)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:112)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.<span class="code-object">Thread</span>.run(<span class="code-object">Thread</span>.java:748)
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> There is also an issue with a simple instantiation of the `groovy.lang.Tuple`. It seems to be blocked by the Script Security. Details are below. Let me know if this is a separate issue and I will raise a different ticket.</p>

<p>Jenkins 2.95<br/>
Script Security 1.39</p>

<p>Test Pipeline</p>


<pre>
node {
    def output = stuff()
    echo output[0]
    echo output[1]
}

def stuff() {
    <span class="code-keyword">return</span> <span class="code-keyword">new</span> Tuple(1, <span class="code-quote">'banana'</span>)
}
</pre>


<pre>
org.jenkinsci.plugins.scriptsecurity.sandbox.RejectedAccessException: unclassified <span class="code-keyword">new</span> groovy.lang.Tuple java.lang.<span class="code-object">Integer</span> java.lang.<span class="code-object">String</span>
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onNewInstance(SandboxInterceptor.java:144)
	at org.kohsuke.groovy.sandbox.impl.Checker$3.call(Checker.java:197)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedConstructor(Checker.java:202)
	at com.cloudbees.groovy.cps.sandbox.SandboxInvoker.constructorCall(SandboxInvoker.java:21)
	at WorkflowScript.stuff(WorkflowScript:8)
	at WorkflowScript.run(WorkflowScript:2)
	at ___cps.transform___(Native Method)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:96)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.fixArg(FunctionCallBlock.java:82)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
	at com.cloudbees.groovy.cps.impl.ConstantBlock.eval(ConstantBlock.java:21)
	at com.cloudbees.groovy.cps.Next.step(Next.java:83)
	at com.cloudbees.groovy.cps.Continuable$1.call(Continuable.java:174)
	at com.cloudbees.groovy.cps.Continuable$1.call(Continuable.java:163)
	at org.codehaus.groovy.runtime.GroovyCategorySupport$ThreadCategoryInfo.use(GroovyCategorySupport.java:122)
	at org.codehaus.groovy.runtime.GroovyCategorySupport.use(GroovyCategorySupport.java:261)
	at com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:163)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.access$001(SandboxContinuable.java:19)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable$1.call(SandboxContinuable.java:35)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable$1.call(SandboxContinuable.java:32)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.GroovySandbox.runInSandbox(GroovySandbox.java:108)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.run0(SandboxContinuable.java:32)
	at org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:174)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:331)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$100(CpsThreadGroup.java:82)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:243)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:231)
	at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:64)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:112)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.<span class="code-object">Thread</span>.run(<span class="code-object">Thread</span>.java:748)
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/duemir.png" title="duemir's avatar" /> duemir:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-45575?focusedId=329496&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-329496">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> It sound's like I don't need to raise a separate <tt>script-security</tt> ticket after all. Correct?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> It sound's like I don't need to raise a separate <tt>script-security</tt> ticket after all. Correct?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-45575?focusedId=331279&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-331279">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Andrew Bayer
Path:
 pom.xml
 src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition2Test.java
<a href="http://jenkins-ci.org/commit/workflow-cps-plugin/9a82584c267f84a4dce925a69d45ad6018196a0a" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/workflow-cps-plugin/9a82584c267f84a4dce925a69d45ad6018196a0a</a>
Log:
  <span class="error">&#91;JENKINS-45575, JENKINS-49679&#93;</span> Test multiple assignment CPS fixes</p>

<p>Specifically <a href="https://github.com/cloudbees/groovy-cps/pull/81" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/cloudbees/groovy-cps/pull/81</a> and <a href="https://github.com/cloudbees/groovy-cps/pull/82" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/cloudbees/groovy-cps/pull/82</a></p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Andrew Bayer<br/>
Path:<br/>
 pom.xml<br/>
 src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition2Test.java<br/>
<a href="http://jenkins-ci.org/commit/workflow-cps-plugin/9a82584c267f84a4dce925a69d45ad6018196a0a" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/workflow-cps-plugin/9a82584c267f84a4dce925a69d45ad6018196a0a</a><br/>
Log:<br/>
  <span class="error">&#91;JENKINS-45575, JENKINS-49679&#93;</span> Test multiple assignment CPS fixes</p>

<p>Specifically <a href="https://github.com/cloudbees/groovy-cps/pull/81" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/cloudbees/groovy-cps/pull/81</a> and <a href="https://github.com/cloudbees/groovy-cps/pull/82" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/cloudbees/groovy-cps/pull/82</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-45575?focusedId=331280&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-331280">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Will be in next release of <tt>workflow-cps</tt> - 2.46.</p>
</pre>
</details>

</ul>
</details>

<p>Will be in next release of <tt>workflow-cps</tt> - 2.46.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-45575?focusedId=331282&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-331282">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Andrew Bayer
Path:
 Jenkinsfile
 pom.xml
 src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition2Test.java
<a href="http://jenkins-ci.org/commit/workflow-cps-plugin/78b137d8743ea3cf80ae302ec89f8aeabf3eb944" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/workflow-cps-plugin/78b137d8743ea3cf80ae302ec89f8aeabf3eb944</a>
Log:
  Merge pull request #206 from abayer/jenkins-45575-jenkins-49679</p>

<p><span class="error">&#91;JENKINS-45575, JENKINS-49679&#93;</span> Test multiple assignment CPS fixes</p>


<p>Compare: <a href="https://github.com/jenkinsci/workflow-cps-plugin/compare/57bda81325d5...78b137d8743e" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/workflow-cps-plugin/compare/57bda81325d5...78b137d8743e</a></p>

</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Andrew Bayer<br/>
Path:<br/>
 Jenkinsfile<br/>
 pom.xml<br/>
 src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition2Test.java<br/>
<a href="http://jenkins-ci.org/commit/workflow-cps-plugin/78b137d8743ea3cf80ae302ec89f8aeabf3eb944" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/workflow-cps-plugin/78b137d8743ea3cf80ae302ec89f8aeabf3eb944</a><br/>
Log:<br/>
  Merge pull request #206 from abayer/jenkins-45575-jenkins-49679</p>

<p><span class="error">&#91;JENKINS-45575, JENKINS-49679&#93;</span> Test multiple assignment CPS fixes</p>


<p>Compare: <a href="https://github.com/jenkinsci/workflow-cps-plugin/compare/57bda81325d5...78b137d8743e" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/workflow-cps-plugin/compare/57bda81325d5...78b137d8743e</a></p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/38846">JENKINS-38846</a>]</i>
